### ISSUE TITLE
For threadsafe implementations, waiting should only require `&self`

### ISSUE THREAD

**Issue by BatmanAoD** [Brendan-Blanchard/async-rate-limit#1]
The `SlidingWindowRateLimiter` is designed in a threadsafe manner, such that its implementation of `wait_until_ready` and `wait_with_cost` could take `&self` instead of `&mut self`. However, since the `RateLimiter` and `VariableCostRateLimiter` traits use `&mut self`, it's not possible for users to take advantage of this.

I suggest either creating a "shared" version of the traits, or just making the `wait...` methods inherent to the struct itself and making `impl <trait>` call those inherent methods.

The "shared" version of `RateLimiter` could be implemented somewhat like this:

```rust
trait SharedRateLimiter {
    fn wait_until_ready(&self) -> impl std::future::Future<Output = ()> + Send;
}

impl<L: SharedRateLimiter> RateLimiter for L {
    fn wait_until_ready(&mut self) -> impl std::future::Future<Output = ()> + Send {
        <Self as SharedRateLimiter>::wait_until_ready(self)
    }
}
```

...and `VariableCostRateLimiter` would be similar.

**Comment by Brendan-Blanchard** [comment]
Hey, thanks for raising the issue! Looks like a nice non-breaking change that adds some utility I haven't had a use case for yet.

Can you confirm that https://github.com/Brendan-Blanchard/async-rate-limit/pull/2 is what you had in mind? I'll add the variable one and get a version out this weekend if so.

**Comment by BatmanAoD** [comment]
Yeah, that's exactly what I need! Thanks!

**Comment by Brendan-Blanchard** [comment]
Perfect! This and the variable version are now released as part of v0.1.1 - please open another issue if something else comes up!
